R/build-articles.R

```diff
@@ -94,31 +112,64 @@ build_articles <- function(pkg = ".", path = "docs/articles", depth = 1L,
 render_rmd <- function(pkg,
                        input,
                        output_file,
+                       rmd_times,
+                       html_times,
                        strip_header = FALSE,
                        data = list(),
                        toc = TRUE,
                        depth = 1L,
                        encoding = "UTF-8",
-                       quiet = TRUE) {
-  message("Building article '", output_file, "'")
+                       quiet = TRUE,
+                       lazy = TRUE) {
+  if (missing(html_times))
+    to_build <- TRUE
+  else {
+    if (!lazy) {
+      to_build <- TRUE
+    }
+    else if (lazy) {
+      if (any(is.na(html_times) || html_times < rmd_times ||
+              html_times == rmd_times)) to_build <- TRUE
+      else if (html_times > rmd_times) to_build <- FALSE
+    }
+  }
 
   format <- build_rmarkdown_format(pkg, depth = depth, data = data, toc = toc)
   on.exit(unlink(format$path), add = TRUE)
 
-  path <- callr::r_safe(
-    function(...) rmarkdown::render(...),
-    args = list(
-      input,
-      output_format = format$format,
-      output_file = basename(output_file),
-      quiet = quiet,
-      encoding = encoding,
-      envir = globalenv()
-    ),
-    show = !quiet
+  arg_list <- list(
+    input,
+    output_format = format$format,
+    output_file = basename(output_file),
+    quiet = quiet,
+    encoding = encoding,
+    envir = globalenv()
   )
-  update_rmarkdown_html(path, strip_header = strip_header, depth = depth,
-    index = pkg$topics)
+
+  if (to_build) {
+      message("Building article '", output_file, "'")
+
+      path <- callr::r_safe(
+        function(...) rmarkdown::render(...),
+        args = arg_list,
+        show = !quiet
+      )
+      update_rmarkdown_html(path, strip_header = strip_header, depth = depth,
+        index = pkg$topics)
+  }
+  else if (!to_build) {
```

Comment:
I want to make sure that the code in the article can always run even if the Rmd file has not changed. This is because the underlying package may have changed in a way that breaks the example.

I attempted to do a different approach with `purl` and `source` but had lots of weird environment and chunk option issues that were hard to test for. It seemed a lot simpler to just build the articles, but not save the results. That way we know that the articles can be fully built without saving lots of non-change changes to the git history.  

That being said, users might want to skip the checking stage altogether to reduce build times. Would it make sense to include an option for that?